test/e2e/framework/metrics/metrics_grabber.go

```diff
@@ -102,6 +109,47 @@ func NewMetricsGrabber(c clientset.Interface, ec clientset.Interface, config *re
 		return nil, errors.New("a rest config is required for grabbing kube-controller and kube-controller-manager metrics")
 	}
 
+	var infos []componentInfo
+	// The component pods might take some time to show up.
+	err := wait.PollImmediate(time.Second*5, time.Minute*5, func() (bool, error) {
```

Comment:
is the logic not the same than the one that already existed to find the pod components but just adding a poll loop?

why not consolidate everything?

also, the comment in the constructor assumes components should be already running

```
// NewMetricsGrabber prepares for grabbing metrics data from several different
// components. It should be called when those components are running because
// it needs to communicate with them to determine for which components
// metrics data can be retrieved.
//
```

Reply:
> is the logic not the same than the one that already existed to find the pod components but just adding a poll loop?

Do you mean `wait.WaitForPodsWithLabel(labels.Everything().String())`? If so, we can use that.

> also, the comment in the constructor assumes components should be already running

According the comment, we should check it before calling it or check/waiting for components running.

Reply:
nope, sorry, after this code, in L153 we have 
```
	podList, err := c.CoreV1().Pods(metav1.NamespaceSystem).List(context.TODO(), metav1.ListOptions{})
	if err != nil {
		return nil, err
	}
	if len(podList.Items) < 1 {
		klog.Warningf("Can't find any pods in namespace %s to grab metrics from", metav1.NamespaceSystem)
	}
	for _, pod := range podList.Items {
		if regKubeScheduler.MatchString(pod.Name) {
			kubeScheduler = pod.Name
		}
		if regKubeControllerManager.MatchString(pod.Name) {
			kubeControllerManager = pod.Name
		}
		if regSnapshotController.MatchString(pod.Name) {
			snapshotControllerManager = pod.Name
		}
		if kubeScheduler != "" && kubeControllerManager != "" && snapshotControllerManager != "" {
			break
		}
	}
```

that seems to be doing the same, just with matchString instead of regex.

However, this is not needed of the caller is respecting the requirements

> ... It should be called when those components are running ...

Why is the caller of the function not verifying this is happening? 
I can't remember issues with this logic in the e2e framework, is this only a problem for e2e_node?